stdlib/mexpr/decision-points.mc

```diff
@@ -0,0 +1,913 @@
+include "mexpr.mc"
+include "pprint.mc"
+include "digraph.mc"
+include "string.mc"
+include "ast-builder.mc"
+include "eq-paths.mc"
+
+-- This file contains implementations related to decision points. In particular,
+-- it implements:
+-- * A language fragment for decision points (HoleAst)
+-- * An algorithm for AST -> call graph conversion (Ast2CallGraph fragment)
+-- * Program transformations for programs with decision points (HolyCallGraph)
+
+-- TODO: this file should probably be split into several ones.
+
+lang HoleAst
+  syn Expr =
+  | TmHole {tp : Type,
+            startGuess : Dyn,
+            depth : Int}
+end
+
+lang HoleAstPrettyPrint = HoleAst + TypePrettyPrint
+  sem pprintCode (indent : Int) =
+  | TmHole h -> strJoin "" ["Hole (", getTypeStringCode indent h.tp, ", ",
+                                      pprintCode indent h.startGuess, ", ",
+                                      pprintCode indent h.depth, ")"]
+end
+
+-- Temporary workaround: uniquely labeled decision points
+lang LHoleAst = HoleAst
+  syn Expr =
+  | LTmHole {tp : Type,
```

Comment:
This should probably be a product extension, once those land

Reply:
Agree.